controllers/cloud.redhat.com/clowdjobinvocation_controller.go

```diff
@@ -152,125 +164,287 @@ func (r *ClowdJobInvocationReconciler) Reconcile(req ctrl.Request) (ctrl.Result,
 		// We have a match that isn't running and can invoke the job
 		r.Log.Info("Invoking job", "jobinvocation", jobName, "namespace", app.Namespace)
 
-		if err := r.InvokeJob(ctx, &job, &app, &env, &cji); err != nil {
-			r.Log.Info("Job Invocation Failed", "jobinvocation", jobName, "namespace", app.Namespace)
+		if err := r.InvokeJob(&cache, &job, &app, &env, &cji); err != nil {
+			r.Log.Error(err, "Job Invocation Failed", "jobinvocation", jobName, "namespace", app.Namespace)
 			r.Recorder.Eventf(&cji, "Warning", "JobNotInvoked", "Job [%s] could not be invoked", jobName)
 			return ctrl.Result{Requeue: true}, err
 		}
 	}
 
+	// Check IQE struct to see if we need to invoke an IQE Job
+	// In the future, we'll need ot handle other types, but this will suffice since testing only has iqe.
+	var emptyTesting crd.IqeJobSpec
+	if cji.Spec.Testing.Iqe != emptyTesting {
+
+		nn := types.NamespacedName{
+			Name:      fmt.Sprintf("%s-iqe", cji.Name),
+			Namespace: cji.Namespace,
+		}
+
+		j := batchv1.Job{}
+
+		if err := r.createIqeJobResource(&cache, &cji, &env, &app, nn, ctx, &j); err != nil {
+			r.Log.Error(err, "Iqe Job creation encountered an error", "jobinvocation", nn.Name)
+			r.Recorder.Eventf(&cji, "Warning", "IQEJobFailure", "Job [%s] failed to invoke", j.ObjectMeta.Name)
+			return ctrl.Result{Requeue: true}, err
+		}
+		if err := cache.Create(IqeClowdJob, nn, &j); err != nil {
```

Comment:
The pattern is that we should create the object in the cache before we modify and then update it afterwards.
```
obj = emptyObject{}
cache.Create(emptyObject...
manipulate(emptyObject)
cache.Update(emptyObject...
```
Whilst in this case it is unlikely the object already exists, the reason for doing it this way is to preserve other settings on the object. If anything else modifies the object, in a way that doesn't conflict with clowder, then clowder will leave those changes alone, and include them in it's "update" otherwise it will just overwrite them. We _may_ wnt this behaviour in some cases, but to keep with consistency I think we should update this for now.

Reply:
Thanks for the clarification here. Will update.